src/lib.rs

```diff
@@ -135,65 +137,51 @@ impl<F> Strawpoll<F> {
 
         let waker = this.waker.as_ref().unwrap();
-        let was_woken = waker.awoken.compare_and_swap(true, false, SeqCst);
+
+        let was_woken = waker
+            .awoken
+            .compare_exchange(true, false, SeqCst, SeqCst)
+            .unwrap_or_else(|f| f);
+
         if !this.was_ready && !was_woken {
             return Poll::Pending;
         }
         this.was_ready = false;
 
-        // safety: we are already pinned, and caller has no way to move us (or F) once we've
+        // SAFETY: we are already pinned, and caller has no way to move us (or F) once we've
         // reached this point unless F: Unpin.
-        let mut fpin = unsafe { Pin::new_unchecked(&mut this.future) };
+        let mut fpin = unsafe { Pin::new_unchecked(&mut this.inner) };
         let wref = futures_task::waker_ref(waker);
         let mut cx = Context::from_waker(&*wref);
-        let mut twice = false;
-        loop {
-            #[cfg(test)]
-            {
-                this.npolls += 1;
-            }
-
-            match poll_fn(fpin.as_mut(), &mut cx) {
-                Poll::Ready(r) => {
-                    // we want to allow polling after a future is ready to support futures that can
-                    // be "reset", like timers. we do this by keeping track of when we return ready
-                    // and bypass the `was_awoken` check the next time `poll_fn` is called.
-                    this.was_ready = true;
-                    return Poll::Ready(r);
-                }
-                Poll::Pending => {
-                    // just in case -- check if we raced and should wake up again immediately
-                    if waker.awoken.compare_and_swap(true, false, SeqCst) == true {
-                        if !twice {
-                            // someone woke us up while we polled -- poll again!
-                            twice = true;
-                            continue;
-                        } else {
-                            // fool me once...
-                            // we've probably just run out of budget and need to yield.
-                            // (https://tokio.rs/blog/2020-04-preemption/)
-                            // we don't need to call .wake() since
-                            // whatever set this to true already did.
-                            waker.awoken.store(true, SeqCst);
-                        }
-                    }
-                    return Poll::Pending;
-                }
-            }
+
+        #[cfg(test)]
+        {
+            this.npolls += 1;
+        }
+
+        let poll = poll_fn(fpin.as_mut(), &mut cx);
+
+        if poll.is_ready() {
+            // we want to allow polling after a future is ready to support futures that can
+            // be "reset", like timers. we do this by keeping track of when we return ready
+            // and bypass the `was_woken` check the next time `poll_fn` is called.
+            this.was_ready = true;
         }
+
```

Comment:
I agree we can remove the "eager retry" logic. It may improve performance marginally in some cases, but it's also not really our responsibility to mitigate such races. I do think we should include a brief comment explaining why a race won't result in a lost wake-up though. (It's because whatever may have raced with us must have also called `.wake()`).

Reply:
Added. Is it enough?

Reply:
I want to tweak it a little but can do that manually after merging!